### PR TITLE
Hoist component statics

### PR DESCRIPTION
**Pull request by leobetosouza**
When we create a React high order component, we must hoist the decorated component static properties and methods. We can see it on React docs: https://reactjs.org/docs/higher-order-components.html#static-methods-must-be-copied-over

The lack of it can generate undesired behavior.

The redux-zero `connect()` is a high order component that don't respect this rule.

On this PR I use the recommended lib to correct it.

### CODE REVIEW

File: src/react/components/connect.tsx [hunk 1]
```diff
@@ -1,9 +1,13 @@
+/* tslint:enable: allowSyntheticDefaultImports */
+
 import * as React from "react";
 
 import shallowEqual from "../../utils/shallowEqual";
 import propValidation from "../../utils/propsValidation";
 import bindActions from "../../utils/bindActions";
 
+const hoistNonReactStatics = require("hoist-non-react-statics");
```
Comment (leobetosouza) [inline src/react/components/connect.tsx]: sadly typescript can't use this module with the `import` statement. :(

**Comment by matheusml** [comment]
Is there a way to do this without the external lib?

**Comment by leobetosouza** [comment]
@matheusml I really really would like to do it without this dependency... but, as you can see here: https://github.com/mridgway/hoist-non-react-statics/blob/master/src/index.js this function do a really specific thing... it is generic and solves the problem... and don't have dependencies.

I can't see now a better way to do it..

**Comment by matheusml** [comment]
Ok, got it. Could you please just update the PR so that the build will propertly pass?

**Comment by leobetosouza** [comment]
@matheusml the Codacy is falling due to `require` statement... i don't know a way to make it pass...

File: src/react/components/connect.tsx (hunk 1 above)
Comment (leobetosouza) [reply src/react/components/connect.tsx]: i've tried it... i don't know sufficient TS to know why this don't works.

On my JS projects I use just `import hoistNonReactStatics from 'hoist-non-react-statics';`

**Comment by Kiho** [comment]
Looks like you need to turn on `"allowSyntheticDefaultImports": true,` in tsconfig.json file.

**Comment by malbernaz** [comment]
`moduleResolution` set to "node" could also do the trick

**Comment by leobetosouza** [comment]
I will try it soon :)

Thanks @Kiho and @malbernaz

**Comment by leobetosouza** [comment]
@Kiho @malbernaz 

Even with ` "moduleResolution": "Node"` and `"allowSyntheticDefaultImports": true` on `compileOptions` I am getting: `TypeError: hoist_non_react_statics_1.default is not a function` when I run tests.

I've tried all combinations of this two options :(

**Comment by Kiho** [comment]
@leobetosouza, @matheusml
I could not find solution to make everybody happy. 
How about this approach?
https://github.com/jaredpalmer/react-fns/blob/3ea8311f5e3c3369fe76aa3ca75e1fe5f01d5905/src/Locales/withLocales.tsx#L10
https://github.com/jaredpalmer/react-fns/blob/3ea8311f5e3c3369fe76aa3ca75e1fe5f01d5905/src/hoistStatics.tsx

**Comment by Kiho** [comment]
Maybe we can use this lib instead https://github.com/mridgway/hoist-non-react-statics

**Comment by matheusml** [comment]
That's very promising @Kiho 

What do you think @leobetosouza ?

**Comment by leobetosouza** [comment]
I dont know if @matheusml would approve this approach...

I prefer the `require` solution... but, if the maintainer of the repo say this is better, i will do.

**Comment by neurosnap** [comment]
You need to install the typedefs for this package: https://github.com/mridgway/hoist-non-react-statics/issues/53

**Comment by matheusml** [comment]
Thanks @leobetosouza 💯

**Comment by makinde** [comment]
Any chance this can be picked up again? It would really be great to have this best practice follows. When using next, components can have `getInitialProps` methods and it's pretty tedious to have to wrap connect them first and then add the static method.

**Pull request merged**